tests/src/Interop/Variant/VariantNative.cpp

```diff
@@ -0,0 +1,349 @@
+
+    return len == TP_SysStringByteLen(expected) && memcmp(value.bstrVal, expected, len) == 0;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Object(VARIANT value)
+{
+    
+    if (value.vt != VT_DISPATCH)
+    {
+        printf("Invalid format. Expected VT_DISPATCH.\n");
+        return FALSE;
+    }
+    
+
+    IDispatch* obj = value.pdispVal;
+
+    if (obj == NULL)
+    {
+        printf("Marshal_ByValue (Native side) recieved an invalid IDispatch pointer\n");
+        return FALSE;
+    }
+
+    obj->AddRef();
+
+    obj->Release();
+
+    return TRUE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Missing(VARIANT value)
+{
+    if (value.vt != VT_ERROR)
+    {
+        printf("Invalid format. Expected VT_ERROR.\n");
+        return FALSE;
+    }
+
+    return value.scode == DISP_E_PARAMNOTFOUND ? TRUE : FALSE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Empty(VARIANT value)
+{
+    if (value.vt != VT_EMPTY)
+    {
+        printf("Invalid format. Expected VT_EMPTY. \n");
+        return FALSE;
+    }
+    
+    return TRUE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Boolean(VARIANT value)
+{
+    if (value.vt != VT_BOOL)
+    {
+        printf("Invalid format. Expected VT_BOOL.\n");
+        return FALSE;
+    }
+
+    return value.boolVal == VARIANT_TRUE ? TRUE : FALSE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_DateTime(VARIANT value)
+{
+    if (value.vt != VT_DATE)
+    {
+        printf("Invalid format. Expected VT_BYREF.\n");
+        return FALSE;
+    }
+
+    BSTR str;
+    
+    // always use the ENGLISH locale so that the string comes out as 11/16/1977 as opposed to 
+    // say 16/11/1977 for German locale; otherwise this test would fail on non-ENU locales
+
+    VarBstrFromDate(value.date, LCID_ENGLISH, VAR_FOURDIGITYEARS, &str);
+
+    if(wcscmp(L"11/6/2018", (wchar_t *)str) != 0 )
+    {
+        wprintf(L"FAILURE! InDATE expected '07/04/2008' but received: %s\n", str);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Decimal(VARIANT value)
+{
+    if (value.vt != VT_DECIMAL)
+    {
+        printf("Invalid format. Expected VT_DECIMAL.\n");
+        return FALSE;
+    }
+
+    return memcmp(&value.decVal, &DecimalValue, sizeof(DECIMAL)) == 0;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Null(VARIANT value)
+{
+    if (value.vt != VT_NULL)
+    {
+        printf("Invalid format. Expected VT_NULL. \n");
+        return FALSE;
+    }
+    
+    return TRUE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByValue_Invalid(VARIANT value)
+{
+    return FALSE;
+}
+
+extern "C" BOOL DLL_EXPORT STDMETHODCALLTYPE Marshal_ByRef(VARIANT* pValue, ExpectedVariantType type)
+{
+    switch (type)
+    {
+#define EXPECTED_VARIANT_TYPE(Type) case Type: return Marshal_ByValue_##Type (*pValue);
```

Comment:
I am sorry, but this is one of my least favorite patterns. the code base is littered with this pattern and with a modern IDE, writing this code is trivial. Please don't use the preprocessor here, just write out the k-statement.